// derby/dist/index.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.util = exports.Derby = exports.use = exports.createApp = exports.ComponentModelData = exports.Component = exports.PageForClient = exports.Page = exports.Dom = exports.App = exports.AppForClient = void 0;
var racer_1 = require("racer");
Object.defineProperty(exports, "util", { enumerable: true, get: function () { return racer_1.util; } });
var Derby_1 = require("./Derby");
Object.defineProperty(exports, "Derby", { enumerable: true, get: function () { return Derby_1.DerbyForClient; } });
var App_1 = require("./App");
Object.defineProperty(exports, "AppForClient", { enumerable: true, get: function () { return App_1.AppForClient; } });
Object.defineProperty(exports, "App", { enumerable: true, get: function () { return App_1.App; } });
var Dom_1 = require("./Dom");
Object.defineProperty(exports, "Dom", { enumerable: true, get: function () { return Dom_1.Dom; } });
var Page_1 = require("./Page");
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return Page_1.Page; } });
Object.defineProperty(exports, "PageForClient", { enumerable: true, get: function () { return Page_1.PageForClient; } });
var components_1 = require("./components");
Object.defineProperty(exports, "Component", { enumerable: true, get: function () { return components_1.Component; } });
Object.defineProperty(exports, "ComponentModelData", { enumerable: true, get: function () { return components_1.ComponentModelData; } });
var DerbyClass = racer_1.util.isServer
    ? racer_1.util.serverRequire(module, './DerbyForServer').DerbyForServer
    : Derby_1.DerbyForClient;
var instance = new DerbyClass();
function createApp(name, file, options) {
    return instance.createApp(name, file, options);
}
exports.createApp = createApp;
function use(plugin, options) {
    return instance.use(plugin, options);
}
exports.use = use;
