// racer/lib/Model/connection.js
Object.defineProperty(exports, "__esModule", { value: true });
var client_1 = require("sharedb/lib/client");
var Model_1 = require("./Model");
var LocalDoc_1 = require("./LocalDoc");
var RemoteDoc_1 = require("./RemoteDoc");
var promisify = require('../util').promisify;
Model_1.Model.INITS.push(function (model) {
    model.root._preventCompose = false;
});
Model_1.Model.prototype.preventCompose = function () {
    var model = this._child();
    model._preventCompose = true;
    return model;
};
Model_1.Model.prototype.allowCompose = function () {
    var model = this._child();
    model._preventCompose = false;
    return model;
};
Model_1.Model.prototype.createConnection = function (bundle) {
    // Model::_createSocket should be defined by the socket plugin
    this.root.socket = this._createSocket(bundle);
    // The Share connection will bind to the socket by defining the onopen,
    // onmessage, etc. methods
    var model = this;
    this.root.connection = new client_1.Connection(this.root.socket);
    this.root.connection.on('state', function (state, reason) {
        model._setDiff(['$connection', 'state'], state);
        model._setDiff(['$connection', 'reason'], reason);
    });
    this._set(['$connection', 'state'], 'connected');
    this._finishCreateConnection();
};
Model_1.Model.prototype._finishCreateConnection = function () {
    var model = this;
    this.root.connection.on('error', function (err) {
        model._emitError(err);
    });
    // Share docs can be created by queries, so we need to register them
    // with Racer as soon as they are created to capture their events
    this.root.connection.on('doc', function (shareDoc) {
        model.getOrCreateDoc(shareDoc.collection, shareDoc.id);
    });
};
Model_1.Model.prototype.connect = function () {
    this.root.socket.open();
};
Model_1.Model.prototype.disconnect = function () {
    this.root.socket.close();
};
Model_1.Model.prototype.reconnect = function () {
    this.disconnect();
    this.connect();
};
// Clean delayed disconnect
Model_1.Model.prototype.close = function (cb) {
    cb = this.wrapCallback(cb);
    var model = this;
    this.whenNothingPending(function () {
        model.root.socket.close();
        cb();
    });
};
Model_1.Model.prototype.closePromised = promisify(Model_1.Model.prototype.close);
// Returns a reference to the ShareDB agent if it is connected directly on the
// server. Will return null if the ShareDB connection has been disconnected or
// if we are not in the same process and we do not have a reference to the
// server-side agent object
Model_1.Model.prototype.getAgent = function () {
    return this.root.connection.agent;
};
Model_1.Model.prototype._isLocal = function (name) {
    // Whether the collection is local or remote is determined by its name.
    // Collections starting with an underscore ('_') are for user-defined local
    // collections, those starting with a dollar sign ('$'') are for
    // framework-defined local collections, and all others are remote.
    var firstCharcter = name.charAt(0);
    return firstCharcter === '_' || firstCharcter === '$';
};
Model_1.Model.prototype._getDocConstructor = function (name) {
    return (this._isLocal(name)) ? LocalDoc_1.LocalDoc : RemoteDoc_1.RemoteDoc;
};
Model_1.Model.prototype.hasPending = function () {
    return this.root.connection.hasPending();
};
Model_1.Model.prototype.hasWritePending = function () {
    return this.root.connection.hasWritePending();
};
Model_1.Model.prototype.whenNothingPending = function (cb) {
    return this.root.connection.whenNothingPending(cb);
};
Model_1.Model.prototype.whenNothingPendingPromised = promisify(Model_1.Model.prototype.whenNothingPending);
