// racer/lib/Model/contexts.js
/**
 * Contexts are useful for keeping track of the origin of subscribes.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Context = exports.Contexts = void 0;
var Model_1 = require("./Model");
var CollectionCounter_1 = require("./CollectionCounter");
Model_1.Model.INITS.push(function (model) {
    model.root._contexts = new Contexts();
    model.root.setContext('root');
});
Model_1.Model.prototype.context = function (contextId) {
    var model = this._child();
    model.setContext(contextId);
    return model;
};
Model_1.Model.prototype.setContext = function (contextId) {
    this._context = this.getOrCreateContext(contextId);
};
Model_1.Model.prototype.getOrCreateContext = function (contextId) {
    var context = this.root._contexts[contextId] ||
        (this.root._contexts[contextId] = new Context(this, contextId));
    return context;
};
Model_1.Model.prototype.unload = function (contextId) {
    var context = (contextId) ? this.root._contexts[contextId] : this._context;
    context && context.unload();
};
Model_1.Model.prototype.unloadAll = function () {
    var contexts = this.root._contexts;
    for (var key in contexts) {
        var currentContext = contexts[key];
        if (contexts.hasOwnProperty(key)) {
            currentContext.unload();
        }
    }
};
var Contexts = /** @class */ (function () {
    function Contexts() {
    }
    Contexts.prototype.toJSON = function () {
        var out = {};
        var contexts = this;
        for (var key in contexts) {
            var currentContext = contexts[key];
            if (currentContext instanceof Context) {
                out[key] = currentContext.toJSON();
            }
        }
        return out;
    };
    ;
    return Contexts;
}());
exports.Contexts = Contexts;
var FetchedQueries = /** @class */ (function () {
    function FetchedQueries() {
    }
    return FetchedQueries;
}());
var SubscribedQueries = /** @class */ (function () {
    function SubscribedQueries() {
    }
    return SubscribedQueries;
}());
var Context = /** @class */ (function () {
    function Context(model, id) {
        this.model = model;
        this.id = id;
        this.fetchedDocs = new CollectionCounter_1.CollectionCounter();
        this.subscribedDocs = new CollectionCounter_1.CollectionCounter();
        this.createdDocs = new CollectionCounter_1.CollectionCounter();
        this.fetchedQueries = new FetchedQueries();
        this.subscribedQueries = new SubscribedQueries();
    }
    Context.prototype.toJSON = function () {
        var fetchedDocs = this.fetchedDocs.toJSON();
        var subscribedDocs = this.subscribedDocs.toJSON();
        var createdDocs = this.createdDocs.toJSON();
        if (!fetchedDocs && !subscribedDocs && !createdDocs)
            return;
        return {
            fetchedDocs: fetchedDocs,
            subscribedDocs: subscribedDocs,
            createdDocs: createdDocs
        };
    };
    ;
    Context.prototype.fetchDoc = function (collectionName, id) {
        this.fetchedDocs.increment(collectionName, id);
    };
    ;
    Context.prototype.subscribeDoc = function (collectionName, id) {
        this.subscribedDocs.increment(collectionName, id);
    };
    ;
    Context.prototype.unfetchDoc = function (collectionName, id) {
        this.fetchedDocs.decrement(collectionName, id);
    };
    ;
    Context.prototype.unsubscribeDoc = function (collectionName, id) {
        this.subscribedDocs.decrement(collectionName, id);
    };
    ;
    Context.prototype.createDoc = function (collectionName, id) {
        this.createdDocs.increment(collectionName, id);
    };
    ;
    Context.prototype.fetchQuery = function (query) {
        mapIncrement(this.fetchedQueries, query.hash);
    };
    ;
    Context.prototype.subscribeQuery = function (query) {
        mapIncrement(this.subscribedQueries, query.hash);
    };
    ;
    Context.prototype.unfetchQuery = function (query) {
        mapDecrement(this.fetchedQueries, query.hash);
    };
    ;
    Context.prototype.unsubscribeQuery = function (query) {
        mapDecrement(this.subscribedQueries, query.hash);
    };
    ;
    Context.prototype.unload = function () {
        var model = this.model;
        for (var hash in this.fetchedQueries) {
            var query = model.root._queries.map[hash];
            if (!query)
                continue;
            var count = this.fetchedQueries[hash];
            while (count--)
                query.unfetch();
        }
        for (var hash in this.subscribedQueries) {
            var query = model.root._queries.map[hash];
            if (!query)
                continue;
            var count = this.subscribedQueries[hash];
            while (count--)
                query.unsubscribe();
        }
        for (var collectionName in this.fetchedDocs.collections) {
            var collection = this.fetchedDocs.collections[collectionName];
            for (var id in collection.counts) {
                var count = collection.counts[id];
                while (count--)
                    model.unfetchDoc(collectionName, id);
            }
        }
        for (var collectionName in this.subscribedDocs.collections) {
            var collection = this.subscribedDocs.collections[collectionName];
            for (var id in collection.counts) {
                var count = collection.counts[id];
                while (count--)
                    model.unsubscribeDoc(collectionName, id);
            }
        }
        for (var collectionName in this.createdDocs.collections) {
            var collection = this.createdDocs.collections[collectionName];
            for (var id in collection.counts) {
                model._maybeUnloadDoc(collectionName, id);
            }
        }
        this.createdDocs.reset();
    };
    ;
    return Context;
}());
exports.Context = Context;
function mapIncrement(map, key) {
    map[key] = (map[key] || 0) + 1;
}
function mapDecrement(map, key) {
    map[key] && map[key]--;
    if (!map[key])
        delete map[key];
}
