// racer/lib/index.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.createBackend = exports.createModel = exports.racer = exports.serverUse = exports.use = exports.RootModel = exports.Racer = exports.util = exports.UnloadEvent = exports.RemoveEvent = exports.MoveEvent = exports.LoadEvent = exports.InsertEvent = exports.ChangeEvent = exports.Context = exports.ModelData = exports.ChildModel = exports.Model = exports.Query = void 0;
var Racer_1 = require("./Racer");
Object.defineProperty(exports, "Racer", { enumerable: true, get: function () { return Racer_1.Racer; } });
var util = require("./util");
var Model_1 = require("./Model");
Object.defineProperty(exports, "RootModel", { enumerable: true, get: function () { return Model_1.RootModel; } });
var Query_1 = require("./Model/Query");
Object.defineProperty(exports, "Query", { enumerable: true, get: function () { return Query_1.Query; } });
var Model_2 = require("./Model");
Object.defineProperty(exports, "Model", { enumerable: true, get: function () { return Model_2.Model; } });
Object.defineProperty(exports, "ChildModel", { enumerable: true, get: function () { return Model_2.ChildModel; } });
Object.defineProperty(exports, "ModelData", { enumerable: true, get: function () { return Model_2.ModelData; } });
var contexts_1 = require("./Model/contexts");
Object.defineProperty(exports, "Context", { enumerable: true, get: function () { return contexts_1.Context; } });
var events_1 = require("./Model/events");
Object.defineProperty(exports, "ChangeEvent", { enumerable: true, get: function () { return events_1.ChangeEvent; } });
Object.defineProperty(exports, "InsertEvent", { enumerable: true, get: function () { return events_1.InsertEvent; } });
Object.defineProperty(exports, "LoadEvent", { enumerable: true, get: function () { return events_1.LoadEvent; } });
Object.defineProperty(exports, "MoveEvent", { enumerable: true, get: function () { return events_1.MoveEvent; } });
Object.defineProperty(exports, "RemoveEvent", { enumerable: true, get: function () { return events_1.RemoveEvent; } });
Object.defineProperty(exports, "UnloadEvent", { enumerable: true, get: function () { return events_1.UnloadEvent; } });
exports.util = require("./util");
var use = util.use, serverUse = util.serverUse;
exports.use = use;
exports.serverUse = serverUse;
exports.racer = new Racer_1.Racer();
/**
 * Creates new RootModel
 *
 * @param data - Optional Data to initialize model with
 * @returns RootModel
 */
function createModel(data) {
    var model = new Model_1.RootModel();
    if (data) {
        model.createConnection(data);
        model.unbundle(data);
    }
    return model;
}
exports.createModel = createModel;
/**
 * Creates racer backend. Can only be called in server process and throws error if called in browser.
 *
 * @param options - Optional
 * @returns racer backend
 */
function createBackend(options) {
    var backendModule = util.serverRequire(module, './Backend');
    if (backendModule == null) {
        throw new Error('racer.createBackend can only be called in server node process');
    }
    return new backendModule.RacerBackend(exports.racer, options);
}
exports.createBackend = createBackend;
